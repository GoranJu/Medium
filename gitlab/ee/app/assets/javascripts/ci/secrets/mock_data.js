import { ROTATION_PERIOD_OPTIONS } from './constants';

/* eslint-disable @gitlab/require-i18n-strings */
const mockSecretsData = [
  {
    name: 'runner-token-windows',
    branch: 'main',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2024-03-11T01:33:06.258Z',
    description: 'This is the runner-token-window secret.',
    environment: 'production',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2026-03-11T01:33:06.258Z',
    labels: ['windows'],
    lastAccessed: '2024-03-19T20:55:08.551Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[0].value,
    status: 'enabled',
  },
  {
    name: 'product/client-tokens',
    branch: 'dev',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2024-01-22T08:04:26.024Z',
    description: null,
    environment: 'staging',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2026-01-22T08:04:26.024Z',
    labels: [],
    lastAccessed: Date.now(),
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: null,
    rotationPeriod: null,
    status: 'disabled',
  },
  {
    name: 'product/server-tokens',
    branch: 'dev-*',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2023-01-24T08:04:26.024Z',
    description: 'This is the product/server-tokens secret.',
    environment: 'staging',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2025-01-24T08:04:26.024Z',
    labels: [],
    lastAccessed: '2024-01-22T07:12:44.833Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[1].value,
    status: 'expiring',
  },
  {
    name: 'security-token-project2',
    branch: 'production',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2024-01-19T11:59:43.516Z',
    description: 'This is the security-token-project2 secret.',
    environment: 'production',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2026-01-19T11:59:43.516Z',
    labels: ['security'],
    lastAccessed: '2024-01-21T23:13:37.227Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[0].value,
    status: 'expired',
  },
  {
    name: 'personal-tokens',
    branch: 'staging',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2024-01-17T05:51:54.602Z',
    description: 'This is the personal-tokens secret.',
    environment: 'development',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2026-01-17T05:51:54.602Z',
    labels: [],
    lastAccessed: '2024-01-19T21:01:38.925Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[2].value,
    status: 'enabled',
  },
  {
    name: 'token-deployment',
    branch: 'staging-*',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2024-01-15T17:12:33.882Z',
    description: 'This is the token-deployment secret.',
    environment: 'none',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2026-01-15T17:12:33.882Z',
    labels: [],
    lastAccessed: '2024-01-16T23:13:31.548Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[2].value,
    status: 'expiring',
  },
  {
    name: 'runner-token-mac',
    branch: 'main',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2023-01-15T16:35:38.326Z',
    description: 'This is the runner-token-mac secret.',
    environment: '*',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2025-01-15T16:35:38.326Z',
    labels: [],
    lastAccessed: '2024-01-06T07:50:58.308Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[0].value,
    status: 'enabled',
  },
  {
    name: 'aws-db-prod-credentials',
    branch: 'dev-*',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2024-01-04T10:37:41.135Z',
    description: 'This is the aws-db-prod-credentials secret.',
    environment: 'none',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2026-01-04T10:37:41.135Z',
    labels: ['AWS'],
    lastAccessed: '2024-01-05T11:47:59.454Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[1].value,
    status: 'disabled',
  },
  {
    name: 'runner-token-linux',
    branch: 'canary',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2023-01-14T18:52:56.258Z',
    description: 'This is the runner-token-linux secret.',
    environment: 'development',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2025-01-14T18:52:56.258Z',
    labels: [],
    lastAccessed: '2023-12-30T16:41:08.551Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[0].value,
    status: 'enabled',
  },
  {
    name: 'SSH-KEY-1',
    branch: 'main',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2023-11-30T01:29:14.394Z',
    description: 'This is the SSH-KEY-1 secret.',
    environment: 'all',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2025-11-30T01:29:14.394Z',
    labels: [],
    lastAccessed: '2023-11-29T22:18:31.562Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[1].value,
    status: 'enabled',
  },
  {
    name: 'group-ID',
    branch: 'staging-*',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2023-01-21T17:34:33.544Z',
    description: 'This is the group-ID secret.',
    environment: 'all',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2025-01-21T17:34:33.544Z',
    labels: ['shell'],
    lastAccessed: '2023-11-29T04:30:10.220Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[1].value,
    status: 'expiring',
  },
  {
    name: 'security-token-project1',
    branch: 'production',
    branchMatchesCount: 2,
    branchMatchesPath: '/branches',
    createdAt: '2023-01-21T17:34:33.544Z',
    description: 'This is the security-token-project1 secret.',
    environment: 'development',
    envMatchesCount: 2,
    envMatchesPath: '/environments',
    expiration: '2025-01-21T17:34:33.544Z',
    labels: ['security'],
    lastAccessed: '2023-11-28T04:30:10.220Z',
    lastAccessedUser: {
      id: 1,
      avatarUrl:
        'https://www.gravatar.com/avatar/83f082bcac69be6bda7945a24ae1a1fda41e864296bd32356819a09cc342e384?s=80&d=identicon',
      name: 'Jane Doe',
      userId: 1,
      username: 'root',
      webUrl: 'http://127.0.0.1:3000/root',
    },
    nextRotation: '2024-09-22T08:04:26.024Z',
    rotationPeriod: ROTATION_PERIOD_OPTIONS[2].value,
    status: 'enabled',
  },
];
/* eslint-enable @gitlab/require-i18n-strings */

export const mockGroupSecretsData = mockSecretsData.map((secret, index) => ({
  ...secret,
  id: index + 1,
}));

export const mockProjectSecretsData = mockSecretsData.map((secret, index) => ({
  ...secret,
  id: index + 1,
}));
