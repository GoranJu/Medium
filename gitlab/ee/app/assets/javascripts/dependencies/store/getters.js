export const isInitialized = ({ currentList, ...state }) => state[currentList].initialized;

export const totals = (state) =>
  state.listTypes.reduce(
    (acc, { namespace }) => ({
      ...acc,
      [namespace]: state[namespace].pageInfo.total || 0,
    }),
    {},
  );
