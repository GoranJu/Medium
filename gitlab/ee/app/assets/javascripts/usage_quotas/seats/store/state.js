export default ({
  namespaceId = null,
  maxFreeNamespaceSeats = null,
  enforcementFreeUserCapEnabled = false,
} = {}) => ({
  initialized: false,
  isLoadingBillableMembers: false,
  isLoadingGitlabSubscription: false,
  isChangingMembershipState: false,
  isRemovingBillableMember: false,
  removedBillableMemberId: null,
  hasError: false,
  namespaceId,
  members: [],
  total: null,
  planCode: null,
  planName: null,
  page: null,
  perPage: null,
  billableMemberToRemove: null,
  userDetails: {},
  search: null,
  sort: 'last_activity_on_desc',
  seatsInSubscription: 0,
  seatsInUse: null,
  maxSeatsUsed: null,
  seatsOwed: null,
  maxFreeNamespaceSeats,
  hasLimitedFreePlan: enforcementFreeUserCapEnabled,
  hasReachedFreePlanLimit: null,
  activeTrial: false,
  subscriptionEndDate: null,
  subscriptionStartDate: null,
});
