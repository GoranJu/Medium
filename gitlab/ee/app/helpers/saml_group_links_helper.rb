# frozen_string_literal: true

module SamlGroupLinksHelper
  def saml_group_link_input_names
    {
      base_access_level_input_name: "saml_group_link[access_level]",
      member_role_id_input_name: "saml_group_link[member_role_id]"
    }
  end

  # For SaaS only. Self-managed configures add-on groups in the configuration file.
  def duo_seat_assignment_available?(group)
    return false unless Feature.enabled?(:saml_groups_duo_add_on_assignment, group)
    return false unless ::Gitlab::Saas.feature_available?(:gitlab_duo_saas_only)
    return false if group.has_parent?

    add_on_purchase = GitlabSubscriptions::Duo.enterprise_or_pro_for_namespace(group)
    return false unless add_on_purchase.present?

    add_on_purchase.active?
  end
end
