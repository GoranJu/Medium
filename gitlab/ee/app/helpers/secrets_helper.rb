# frozen_string_literal: true

module SecretsHelper
  def project_secrets_app_data(project)
    {
      project_path: project.full_path,
      project_secrets_settings_path: project_settings_ci_cd_path(
        project, expand_secrets: true, anchor: 'js-secrets-settings'
      ),
      project_id: project.id,
      base_path: project_secrets_path(project)
    }
  end
end
