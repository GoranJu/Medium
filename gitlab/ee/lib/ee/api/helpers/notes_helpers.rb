# frozen_string_literal: true

module EE
  module API
    module Helpers
      module NotesHelpers
        extend ActiveSupport::Concern

        class_methods do
          extend ::Gitlab::Utils::Override

          override :feature_category_per_noteable_type
          def feature_category_per_noteable_type
            super.merge!(
              ::Epic => :portfolio_management,
              ::Vulnerability => :vulnerability_management
            )
          end
        end

        def add_parent_to_finder_params(finder_params, noteable_type)
          if noteable_type.name.underscore == 'epic'
            finder_params[:group_id] = user_group.id
          else
            super
          end
        end

        # private projects can be found as long as user is a project member.
        # internal projects can be found by any authenticated user.
        def find_merge_request(merge_request_iid)
          params = finder_params_by_noteable_type_and_id(::MergeRequest, merge_request_iid)

          ::NotesFinder.new(current_user, params).target || not_found!(::MergeRequest)
        end
      end
    end
  end
end
