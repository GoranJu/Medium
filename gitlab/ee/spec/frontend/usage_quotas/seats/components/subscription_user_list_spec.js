import {
  GlPagination,
  GlButton,
  GlTable,
  GlAvatarLink,
  GlAvatarLabeled,
  GlBadge,
  GlModal,
  GlTooltip,
} from '@gitlab/ui';
import { mount, shallowMount } from '@vue/test-utils';
import Vue, { nextTick } from 'vue';
// eslint-disable-next-line no-restricted-imports
import Vuex from 'vuex';
import SubscriptionUserList, {
  FIVE_MINUTES_IN_MS,
} from 'ee/usage_quotas/seats/components/subscription_user_list.vue';
import {
  CANNOT_REMOVE_BILLABLE_MEMBER_MODAL_CONTENT,
  DELETED_BILLABLE_MEMBERS_EXPIRES_STORAGE_KEY_SUFFIX,
  DELETED_BILLABLE_MEMBERS_STORAGE_KEY_SUFFIX,
  SORT_OPTIONS,
} from 'ee/usage_quotas/seats/constants';
import getStoreConfig from 'ee/usage_quotas/seats/store';
import { REMOVE_BILLABLE_MEMBER_SUCCESS } from 'ee/usage_quotas/seats/store/mutation_types';
import { mockTableItems } from 'ee_jest/usage_quotas/seats/mock_data';
import { useLocalStorageSpy } from 'helpers/local_storage_helper';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import SearchAndSortBar from '~/usage_quotas/components/search_and_sort_bar/search_and_sort_bar.vue';
import { useFakeDate } from 'helpers/fake_date';

Vue.use(Vuex);

const MOCK_SEAT_USAGE_EXPORT_PATH = '/groups/test_group/-/seat_usage.csv';

const actionSpies = {
  setBillableMemberToRemove: jest.fn(),
  setSearchQuery: jest.fn(),
};

const defaultProvide = {
  subscriptionHistoryHref: '/groups/my-group/-/usage_quotas/subscription_history.csv',
  glFeatures: {
    billableMemberAsyncDeletion: true,
  },
  seatUsageExportPath: MOCK_SEAT_USAGE_EXPORT_PATH,
};

const defaultProps = {
  hasFreePlan: false,
};

const fakeStore = ({ initialState, initialGetters }) =>
  new Vuex.Store({
    actions: actionSpies,
    mutations: getStoreConfig().mutations,
    getters: {
      tableItems: () => mockTableItems,
      isLoading: () => false,
      ...initialGetters,
    },
    state: {
      hasError: false,
      namespaceId: 1,
      total: 300,
      page: 1,
      perPage: 5,
      sort: 'last_activity_on_desc',
      removedBillableMemberId: null,
      ...initialState,
    },
  });

describe('SubscriptionUserList', () => {
  useLocalStorageSpy();
  useFakeDate('2025-03-16T15:00:00.000Z');

  let store;
  let wrapper;

  const localStorageKey = `13-${DELETED_BILLABLE_MEMBERS_STORAGE_KEY_SUFFIX}`;
  const localStorageExpireKey = `13-${DELETED_BILLABLE_MEMBERS_EXPIRES_STORAGE_KEY_SUFFIX}`;
  const fiveMinutesBeforeNow = () => new Date().getTime() - FIVE_MINUTES_IN_MS;
  const fiveMinutesFromNow = () => new Date().getTime() + FIVE_MINUTES_IN_MS;

  const createComponent = ({
    mountFn = shallowMount,
    initialState = {},
    initialGetters = {},
    provide = {},
    props = {},
  } = {}) => {
    store = fakeStore({ initialGetters, initialState });
    wrapper = extendedWrapper(
      mountFn(SubscriptionUserList, {
        store,
        provide: {
          ...defaultProvide,
          ...provide,
        },
        propsData: {
          ...defaultProps,
          ...props,
        },
        stubs: {
          SearchAndSortBar: true,
        },
      }),
    );
  };

  const findTable = () => wrapper.findComponent(GlTable);
  const findExportButton = () => wrapper.findByTestId('export-button');
  const findExportSeatUsageHistoryButton = () =>
    wrapper.findByTestId('subscription-seat-usage-history');
  const findSearchAndSortBar = () => wrapper.findComponent(SearchAndSortBar);
  const findPagination = () => wrapper.findComponent(GlPagination);
  const findAllRemoveUserItems = () => wrapper.findAllByTestId('remove-user');
  const findRemoveMemberItem = (id) => wrapper.find(`[id="remove-member-${id}"]`);
  const findErrorModal = () => wrapper.findComponent(GlModal);

  const serializeTableRow = (rowWrapper) => {
    const emailWrapper = rowWrapper.find('[data-testid="email"]');

    return {
      email: emailWrapper.text(),
      tooltip: emailWrapper.find('span').attributes('title'),
      removeUserButtonExists: rowWrapper.findComponent(GlButton).exists(),
      lastActivityOn: rowWrapper.find('[data-testid="last_activity_on"]').text(),
      lastLoginAt: rowWrapper.find('[data-testid="last_login_at"]').text(),
    };
  };

  const findSerializedTable = (tableWrapper) => {
    return tableWrapper.findAll('tbody tr').wrappers.map(serializeTableRow);
  };

  afterEach(() => {
    localStorage.clear();
  });

  describe('renders', () => {
    beforeEach(() => {
      createComponent({
        mountFn: mount,
        initialGetters: {
          tableItems: () => mockTableItems,
        },
      });
    });

    describe('export button', () => {
      it('has the correct href', () => {
        expect(findExportButton().attributes().href).toBe(MOCK_SEAT_USAGE_EXPORT_PATH);
      });
    });

    describe('ExportSeatUsageHistoryButton', () => {
      it('has the correct href', () => {
        expect(findExportSeatUsageHistoryButton().attributes().href).toBe(
          defaultProvide.subscriptionHistoryHref,
        );
      });

      describe('with a Free Plan', () => {
        beforeEach(() => {
          createComponent({
            mountFn: mount,
            initialGetters: {
              tableItems: () => mockTableItems,
            },
            props: {
              hasFreePlan: true,
            },
          });
        });

        it('does not render if plan is free', () => {
          expect(findExportSeatUsageHistoryButton().exists()).toBe(false);
        });
      });
    });

    describe('table content', () => {
      it('renders the correct data', () => {
        const serializedTable = findSerializedTable(findTable());

        expect(serializedTable).toMatchSnapshot();
      });
    });

    it('pagination is rendered and passed correct values', () => {
      const pagination = findPagination();

      expect(pagination.props()).toMatchObject({
        perPage: 5,
        totalItems: 300,
      });
    });

    describe('with error modal', () => {
      it('does not render the modal if the user is not removable', async () => {
        await findAllRemoveUserItems().at(0).trigger('click');

        expect(findErrorModal().html()).toBe('');
      });

      it('renders the error modal if the user is removable', async () => {
        await findAllRemoveUserItems().at(2).trigger('click');

        expect(findErrorModal().text()).toContain(CANNOT_REMOVE_BILLABLE_MEMBER_MODAL_CONTENT);
      });
    });

    describe('when removing a billable user', () => {
      const [{ user }] = mockTableItems;

      describe('with billableMemberAsyncDeletion enabled', () => {
        beforeEach(() => {
          createComponent({ initialState: { namespaceId: 13 } });
          return store.commit(REMOVE_BILLABLE_MEMBER_SUCCESS, { memberId: user.id });
        });

        it('sets the local storage key for the member id', () => {
          expect(localStorage.setItem).toHaveBeenCalledWith(localStorageKey, `[${user.id}]`);
        });

        it('sets the local storage key for expiration', () => {
          expect(localStorage.setItem).toHaveBeenCalledWith(
            localStorageExpireKey,
            fiveMinutesFromNow(),
          );
        });

        describe('when removing another member', () => {
          it('sets the local storage key for the member id', async () => {
            store.commit(REMOVE_BILLABLE_MEMBER_SUCCESS, { memberId: 13 });

            await nextTick();

            expect(localStorage.setItem).toHaveBeenLastCalledWith(
              localStorageKey,
              `[${user.id},13]`,
            );
          });

          it('sets the local storage key for expiration', () => {
            expect(localStorage.setItem).toHaveBeenCalledWith(
              localStorageExpireKey,
              fiveMinutesFromNow(),
            );
          });
        });
      });

      describe('with billableMemberAsyncDeletion disabled', () => {
        beforeEach(() => {
          createComponent({
            initialState: { namespaceId: 13 },
            provide: {
              glFeatures: {
                billableMemberAsyncDeletion: false,
              },
            },
          });
          return store.commit(REMOVE_BILLABLE_MEMBER_SUCCESS, { memberId: user.id });
        });

        it('does not set items in the local storage', () => {
          expect(localStorage.setItem).not.toHaveBeenCalled();
        });
      });
    });

    describe('when the removed billable user is set', () => {
      const selectedItem = 0;
      const { user } = mockTableItems[selectedItem];

      beforeEach(() => {
        createComponent({ initialState: { removedBillableMemberId: user.id }, mountFn: mount });
      });

      it('disables the related remove button', () => {
        expect(findAllRemoveUserItems().at(selectedItem).attributes().disabled).toBe('disabled');
      });

      it('does not disable unrelated remove button', () => {
        expect(findAllRemoveUserItems().at(1).attributes().disabled).toBeUndefined();
      });

      it('shows a tooltip for related users', () => {
        expect(findRemoveMemberItem(user.id).findComponent(GlTooltip).text()).toBe(
          'This user is scheduled for removal.',
        );
      });

      it('does snot show a tooltip for unrelated user', () => {
        const [, { user: nonRemovedUser }] = mockTableItems;

        expect(findRemoveMemberItem(nonRemovedUser.id).findComponent(GlTooltip).exists()).toBe(
          false,
        );
      });

      describe('with billableMemberAsyncDeletion disabled', () => {
        beforeEach(() => {
          createComponent({
            initialState: { namespaceId: 13 },
            mountFn: mount,
            provide: {
              glFeatures: {
                billableMemberAsyncDeletion: false,
              },
            },
          });
        });

        it('does not disable the related remove button', () => {
          expect(findAllRemoveUserItems().at(selectedItem).attributes().disabled).toBeUndefined();
        });
      });
    });

    describe('when the removed billable user is in local storage', () => {
      const selectedItem = 0;
      const { user } = mockTableItems[selectedItem];

      beforeEach(() => {
        localStorage.setItem(localStorageKey, `[${user.id}]`);
        localStorage.setItem(localStorageExpireKey, fiveMinutesFromNow());
        createComponent({ initialState: { namespaceId: 13 }, mountFn: mount });
      });

      it('disables the related remove button', () => {
        expect(findAllRemoveUserItems().at(selectedItem).attributes().disabled).toBe('disabled');
      });

      it('does not disable unrelated remove button', () => {
        expect(findAllRemoveUserItems().at(1).attributes().disabled).toBeUndefined();
      });

      it('shows a tooltip for related users', () => {
        expect(findRemoveMemberItem(user.id).findComponent(GlTooltip).text()).toBe(
          'This user is scheduled for removal.',
        );
      });

      it('does snot show a tooltip for unrelated user', () => {
        const [, { user: nonRemovedUser }] = mockTableItems;

        expect(findRemoveMemberItem(nonRemovedUser.id).findComponent(GlTooltip).exists()).toBe(
          false,
        );
      });

      describe('when the local storage item is expired', () => {
        beforeEach(() => {
          localStorage.setItem(localStorageKey, `[${user.id}]`);
          localStorage.setItem(localStorageExpireKey, fiveMinutesBeforeNow());
          createComponent({ initialState: { namespaceId: 13 }, mountFn: mount });
        });

        it('does not disable the related remove button', () => {
          expect(findAllRemoveUserItems().at(selectedItem).attributes().disabled).toBeUndefined();
        });

        it('removes the local storage key', () => {
          expect(localStorage.removeItem).toHaveBeenCalledWith(localStorageKey);
        });
      });

      describe('when the local storage item does not match the user', () => {
        it('does not disable the related remove button', () => {
          localStorage.setItem(localStorageKey, `[11]`);
          localStorage.setItem(localStorageExpireKey, fiveMinutesFromNow());
          createComponent({ initialState: { namespaceId: 13 }, mountFn: mount });

          expect(findAllRemoveUserItems().at(selectedItem).attributes().disabled).toBeUndefined();
        });
      });

      describe('when the local storage item does not match namespace id', () => {
        it('does not disable the related remove button', () => {
          localStorage.setItem(localStorageKey, `[${user.id}]`);
          localStorage.setItem(localStorageExpireKey, fiveMinutesFromNow());
          createComponent({ initialState: { namespaceId: 11 }, mountFn: mount });

          expect(findAllRemoveUserItems().at(selectedItem).attributes().disabled).toBeUndefined();
        });
      });

      describe('when the local storage errors', () => {
        it('does not disable any remove button', () => {
          localStorage.setItem.mockImplementation(() => {
            throw new Error('This is an error');
          });
          createComponent({ initialState: { namespaceId: 11 }, mountFn: mount });

          expect(findAllRemoveUserItems().at(selectedItem).attributes().disabled).toBeUndefined();
        });
      });
    });

    describe('members labeled avatar', () => {
      it('shows the correct avatarLabeled length', () => {
        const avatarLabeledList = findTable().findAllComponents(GlAvatarLabeled);

        expect(avatarLabeledList).toHaveLength(6);
      });

      it('passes the correct props to avatarLabeled', () => {
        const avatarLabeled = findTable().findComponent(GlAvatarLabeled);

        expect(avatarLabeled.props()).toMatchObject({ label: 'Administrator', subLabel: '@root' });
      });
    });

    describe('members avatar', () => {
      it('shows the correct avatarLinks length', () => {
        const avatarLinks = findTable().findAllComponents(GlAvatarLink);

        expect(avatarLinks).toHaveLength(6);
      });

      it('passes the correct props to avatarLink', () => {
        const avatarLink = findTable().findComponent(GlAvatarLink);

        expect(avatarLink.attributes()).toMatchObject({
          alt: 'Administrator',
          href: 'path/to/administrator',
        });
      });

      it.each(['group_invite', 'project_invite'])(
        'shows the correct badge for membership_type %s',
        (membershipType) => {
          const avatarLinks = findTable().findAllComponents(GlAvatarLink);
          const badgeText = (
            membershipType.charAt(0).toUpperCase() + membershipType.slice(1)
          ).replace('_', ' ');

          avatarLinks.wrappers.forEach((avatarLinkWrapper) => {
            const currentMember = mockTableItems.find(
              (item) => item.user.name === avatarLinkWrapper.attributes().alt,
            );

            if (membershipType === currentMember.user.membership_type) {
              expect(avatarLinkWrapper.findComponent(GlBadge).text()).toBe(badgeText);
            }
          });
        },
      );
    });

    describe('members details always shown', () => {
      it.each`
        membershipType
        ${'project_invite'}
        ${'group_invite'}
        ${'project_member'}
        ${'group_member'}
      `(
        'when membershipType is $membershipType, shouldShowDetails will be true',
        ({ membershipType }) => {
          mockTableItems.forEach((item) => {
            const detailsExpandButtons = findTable().find(
              `[data-testid="toggle-seat-usage-details-${item.user.id}"]`,
            );

            if (membershipType === item.user.membership_type) {
              expect(detailsExpandButtons.exists()).toBe(true);
            }
          });
        },
      );
    });
  });

  describe('Loading state', () => {
    describe('when nothing is loading', () => {
      beforeEach(() => {
        createComponent();
      });

      it('displays the table in a non-busy state', () => {
        expect(findTable().attributes('busy')).toBe(undefined);
      });
    });

    describe.each([
      [true, false],
      [false, true],
    ])('busy when isLoading=%s and hasError=%s', (isLoading, hasError) => {
      beforeEach(() => {
        createComponent({
          initialGetters: { isLoading: () => isLoading },
          initialState: { hasError },
        });
      });

      it('displays table in busy state', () => {
        expect(findTable().attributes('busy')).toBe('true');
      });
    });
  });

  describe('search box', () => {
    beforeEach(() => {
      createComponent();
    });

    it('input event triggers the setSearchQuery action', () => {
      const SEARCH_STRING = 'search string';

      findSearchAndSortBar().vm.$emit('onFilter', SEARCH_STRING);

      expect(actionSpies.setSearchQuery).toHaveBeenCalledTimes(1);
      expect(actionSpies.setSearchQuery).toHaveBeenCalledWith(expect.any(Object), SEARCH_STRING);
    });

    it('contains the correct sort options', () => {
      expect(findSearchAndSortBar().props('sortOptions')).toMatchObject(SORT_OPTIONS);
    });
  });
});
