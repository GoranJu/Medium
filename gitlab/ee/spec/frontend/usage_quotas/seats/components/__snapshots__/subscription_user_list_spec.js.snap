// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionUserList renders table content renders the correct data 1`] = `
[
  {
    "email": "administrator@email.com",
    "lastActivityOn": "2020-03-01",
    "lastLoginAt": "2022-11-10 10:58:05",
    "removeUserButtonExists": true,
    "tooltip": undefined,
  },
  {
    "email": "agustin_walker@email.com",
    "lastActivityOn": "2020-03-01",
    "lastLoginAt": "2021-01-20 10:58:05",
    "removeUserButtonExists": true,
    "tooltip": undefined,
  },
  {
    "email": "Private",
    "lastActivityOn": "Never",
    "lastLoginAt": "Never",
    "removeUserButtonExists": true,
    "tooltip": "An email address is only visible for users with public emails.",
  },
  {
    "email": "jdoe@email.com",
    "lastActivityOn": "Never",
    "lastLoginAt": "Never",
    "removeUserButtonExists": true,
    "tooltip": undefined,
  },
  {
    "email": "jsnow@email.com",
    "lastActivityOn": "2020-03-01",
    "lastLoginAt": "Never",
    "removeUserButtonExists": true,
    "tooltip": undefined,
  },
  {
    "email": "current_user@email.com",
    "lastActivityOn": "2020-03-01",
    "lastLoginAt": "Never",
    "removeUserButtonExists": true,
    "tooltip": undefined,
  },
]
`;
